### PR TITLE
gh-101578: Amend PyErr_{Set,Get}RaisedException docs

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-101578

### CODE REVIEW

**Comment by erlend-aasland** [comment]
- I removed the duplicate example; it should be sufficient to provide a single example.
- Use `note::` to put emphasis on non-standard reference count semantics in the set function.
- Mark the getter as returning a borrowed reference.
- Apply Diátaxis: be succinct and to the point (and hopefully also clearer)

File: Doc/c-api/exceptions.rst
Comment (erlend-aasland) [inline Doc/c-api/exceptions.rst]: Perhaps this sentence should live in the `seealso` directive.

File: Doc/c-api/exceptions.rst
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]:
```suggestion
   For example::
```

Since this is prose

File: Doc/c-api/exceptions.rst
Comment (CAM-Gerlach) [reply Doc/c-api/exceptions.rst]: Yes; it would make a good bit more sense IMO than the current, since it seems rather redundant to me to have a whole See Also callout for the very next function that just one line away, which to me is more confusing and distracting than helpful. I've made a suggestion to that effect.

File: Doc/c-api/exceptions.rst
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]:
```suggestion
   .. seealso:: :c:func:`PyErr_GetHandledException`, to fetch the exception currently being handled.
```

As discussed above

File: Doc/c-api/exceptions.rst
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]:
```suggestion
   .. caution::
```

Maybe worth using an admonition class more appropriate for an advisory message, rather than a mere informational one—i.e. bad things will happen if you don't pay attention to this, which is the intended semantics of `caution`.

File: Doc/c-api/exceptions.rst
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]:
```suggestion
   .. seealso:: :c:func:`PyErr_GetRaisedException`, to save the current exception.
```

I find it helpful here (and generally) to provide a bit of explicit context as to _why_ readers might want to see this other function.

**Review by CAM-Gerlach — COMMENTED** [review]
Thanks @erlend-aasland ; this is a substantial improvement to the previous docs.

Perhaps its worth fixing the other couple remaining similar docs defects also introduced in PR #101607 , namely in the descriptions of the newly documented functions `PyException_GetArgs` and `PyException_SetArgs` (that could use a similar (if much simpler) treatment as here), and the deprecation notices added to `PyErr_Fetch`, `PyErr_Restore` and `PyErr_NormalizeException` not being added to, and particularly duplicating, the `.. deprecated` directive.

File: Doc/c-api/exceptions.rst
Comment (erlend-aasland) [reply Doc/c-api/exceptions.rst]: TIL!

**Comment by erlend-aasland** [comment]
> Thanks @erlend-aasland ; this is a substantial improvement to the previous docs.

Thanks, and thanks for the review!

> Perhaps its worth fixing the other couple remaining similar docs defects also introduced in PR #101607 , namely in the descriptions of the newly documented functions `PyException_GetArgs` and `PyException_SetArgs` (that could use a similar (if much simpler) treatment as here), and the deprecation notices added to `PyErr_Fetch`, `PyErr_Restore` and `PyErr_NormalizeException` not being added to, and particularly duplicating, the `.. deprecated` directive.

Definitely; perhaps as a follow-up PR?

**Comment by CAM-Gerlach** [comment]
> Definitely; perhaps as a follow-up PR?

Yeah certainly could be; up to you, of course

File: Doc/c-api/exceptions.rst
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]: As mentioned previously, I would suggest eliminating this one as it is listed on the very next line—it would be rather noisy to list the many directly adjacent related functions like this.

```suggestion
   .. seealso:: :c:func:`PyErr_GetHandledException`,
                to fetch the exception currently being handled.
```

Or, if it really does need both, as [officially documented](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-seealso) and done elsewhere, a See-Also should have its contents formatted as a definition list:

```suggestion
   .. seealso::

      :c:func:`PyErr_SetRaisedException`
         To set the current raised exception.

      :c:func:`PyErr_GetHandledException`
         To fetch the exception currently being handled.
```

File: Doc/c-api/exceptions.rst
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]:
```suggestion
```

Either way, eliminate this (as I've incorporated is as necessary above)

File: Doc/c-api/exceptions.rst
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]: Unfortunately, the existing Python docs theme appears to be completely missing formatting for the caution admonition—in fact, it appears to lack such for every admonition I tested except for `warning` and `note`. So, it seems we have to either have to upgrade this to a warning:

```suggestion
   .. warning::
```

or, downgrade it back to a note:

```suggestion
   .. note::
```

**Review by CAM-Gerlach — COMMENTED** [review]
A couple followup suggestions

File: Doc/c-api/exceptions.rst
Comment (erlend-aasland) [reply Doc/c-api/exceptions.rst]: I'd prefer a warning. APIs that steal references are a source of bugs; IMO it is worth it to place emphasis on such deviances.

File: Doc/c-api/exceptions.rst
Comment (erlend-aasland) [reply Doc/c-api/exceptions.rst]: I'm not sure if we need the cross-linking between the set/get APIs. IMO, it leads to too much clutter.

**Review by CAM-Gerlach — APPROVED** [review]
LGTM from me; thanks @erlend-aasland !

**Comment by erlend-aasland** [comment]
Thanks for the review, CAM; highly appreciated. I'll wait for @iritkatriel's or @markshannon's approval before landing.

File: Doc/c-api/exceptions.rst
Comment (iritkatriel) [inline Doc/c-api/exceptions.rst]: I think this is a little more vague than "If the exception is already set" (to understand this I need to know that the two mean the same thing).

Comment (CAM-Gerlach) [reply Doc/c-api/exceptions.rst]: Are any of these suggestions clearer for you?

```suggestion
   If an exception is already being raised, it is cleared first.
```

or

```suggestion
   If an exception is already set as being raised, it is cleared first.
```

or

```suggestion
   If an exception is already set, that exception is cleared first.
```

Comment (iritkatriel) [inline Doc/c-api/exceptions.rst]:
```suggestion
   Clear the exception currently being raised and set it to *exc*.
```

Comment (CAM-Gerlach) [reply Doc/c-api/exceptions.rst]: I like the more concise language, though the above implies that another exception will always be raised when this is used, which I'm guessing is not necessarily the case. Refining it further to avoid that, leading with the parameter name for easy scannability and consistency of reference docs per Diataxis, and making the sentence structure a little easier to parse:

```suggestion
   Set *exc* as the exception currently being raised,
   clearing the existing exception if one is set.
```

Comment (CAM-Gerlach) [reply Doc/c-api/exceptions.rst]: Presumably continued [here](https://github.com/python/cpython/pull/101962#discussion_r1110978468)

Comment (iritkatriel) [reply Doc/c-api/exceptions.rst]: LGTM

**Comment by erlend-aasland** [comment]
Thanks for the reviews, Irit and CAM; highly appreciated. I'll create a follow-up PR Any Time Soon™.

**Pull request merged**
